src/Microsoft.VisualStudio.Threading/JoinableTaskDependencyGraph.cs

```diff
@@ -667,7 +667,7 @@ internal void OnTaskCompleted(IJoinableTaskDependent thisDependentNode)
                         var childrenTasks = new List<IJoinableTaskDependent>(this.childDependentNodes.Keys);
                         while (existingTaskTracking is object)
                         {
-                            RemoveDependingSynchronousTaskFrom(childrenTasks, existingTaskTracking.SynchronousTask, false);
+                            RemoveDependingSynchronousTaskFrom(childrenTasks, existingTaskTracking.SynchronousTask, force: existingTaskTracking.SynchronousTask == thisDependentNode);
```

Comment:
*This* looks like a functional change. Is it intentional?

Reply:
yes, it is intentional. This code is responsible to remove outgoing dependency tracking when a JTF is completed. It happens to be called in when a main thread blocking synchronous task is completed itself. Those tasks have another clean-up code before CompleteOnCurrentThread exits. It appears to be a race condition which one runs first.

 When synchronous task is completed, there is no reason to track ref-count to itself, which is done inside CompleteOnCurrentThread side by turning on this flag. But we track refcount here. Keep ref-count can be more expensive as we add nodes to the PotentialDisconnectedItems collection and clean them up when ref-count goes to 0. It doesn't impact the function, because anything left in the collection will be cleaned up by the second call, but it is basically an extra overhead.